### PR TITLE
UCP/EP/CM: fix close protocol + error handling

### PR DESCRIPTION
**Pull request by evgeny-leksikov**
covered cases:
 - err callback is invoked more than once
 - incorrect handling of disconnect event on failed endpoint
 - disconnecting of not connected CM lane on server side
 - wrong ucp_cm_cbs_remove_filter

### CODE REVIEW

File: src/ucp/core/ucp_ep.c [hunk 1]
```diff
@@ -2097,6 +2097,12 @@ uint64_t ucp_ep_get_tl_bitmap(ucp_ep_h ep)
 
 void ucp_ep_invoke_err_cb(ucp_ep_h ep, ucs_status_t status)
 {
+    if (ep->flags & UCP_EP_FLAG_INVALIDATED) {
+        return;
+    }
+
+    ep->flags |= UCP_EP_FLAG_INVALIDATED;
```
Comment (yosefe) [inline src/ucp/core/ucp_ep.c:2104]: maybe 
UCP_EP_FLAG_ERR_HANDLER_INVOKED

File: src/ucp/core/ucp_ep.c [hunk 2]
```diff
@@ -2097,6 +2097,12 @@ uint64_t ucp_ep_get_tl_bitmap(ucp_ep_h ep)
 
 void ucp_ep_invoke_err_cb(ucp_ep_h ep, ucs_status_t status)
 {
+    if (ep->flags & UCP_EP_FLAG_INVALIDATED) {
```
Comment (alinask) [inline src/ucp/core/ucp_ep.c:2100]: can move this to the 'if' below?

File: src/ucp/core/ucp_ep.c (hunk 1 above)
Comment (evgeny-leksikov) [reply src/ucp/core/ucp_ep.c:2104]: ok, will rename

File: src/ucp/core/ucp_ep.c (hunk 2 above)
Comment (evgeny-leksikov) [reply src/ucp/core/ucp_ep.c:2100]: can move but then need to set the flag in 2 places and long if harder to read, i would prefer to handle it separately, + most likely we can remove this case after refactoring of err handling in UCT

**Comment by evgeny-leksikov** [comment]
https://github.com/openucx/ucx/pull/5075/commits/805f78d1780770aff0f7917249ebedf99750c97d and https://github.com/openucx/ucx/pull/5075/commits/c0826d192c62600696c095f8ad19dde1ff7a691b are cherry-picked from https://github.com/openucx/ucx/pull/5058

**Comment by alinask** [comment]
the failure is from a new test in this PR:

```
[2020-04-24T23:02:19.288Z] [ RUN      ] shm_ib/test_ucp_sockaddr_destroy_ep_on_err.onesided_client_cforce/3 <shm,ib>
[2020-04-24T23:02:19.288Z] [     INFO ] Testing fe80::f652:14ff:fe01:b892:0
[2020-04-24T23:02:19.288Z] [     INFO ] server listening on fe80::f652:14ff:fe01:b892:43694
[2020-04-24T23:02:19.850Z] [boo31:23906:0:23906]  ucp_worker.h:312  Assertion `ep->worker == worker' failed: worker=0x1a8e2f0 ep=0x7f5afac94000 ep->worker=0x7f5afac94070
[2020-04-24T23:02:20.777Z] 
[2020-04-24T23:02:20.777Z] /scrap/jenkins/workspace/ucx-3/contrib/../src/ucp/core/ucp_worker.h: [ ucp_worker_get_ep_by_ptr() ]
[2020-04-24T23:02:20.777Z]       ...
[2020-04-24T23:02:20.777Z]       308 
[2020-04-24T23:02:20.777Z]       309     ucs_assert(ep != NULL);
[2020-04-24T23:02:20.777Z]       310     ucs_assertv(ep->worker == worker, "worker=%p ep=%p ep->worker=%p", worker,
[2020-04-24T23:02:20.777Z] ==>   311                 ep, ep->worker);
[2020-04-24T23:02:20.777Z]       312     return ep;
[2020-04-24T23:02:20.777Z]       313 }
[2020-04-24T23:02:20.777Z]       314 
[2020-04-24T23:02:20.777Z] 
[2020-04-24T23:02:21.704Z] ==== backtrace (tid:  23906) ====
[2020-04-24T23:02:21.704Z]  0 0x000000000006fe05 ucp_worker_get_ep_by_ptr()  /scrap/jenkins/workspace/ucx-3/contrib/../src/ucp/core/ucp_worker.h:311
[2020-04-24T23:02:21.704Z]  1 0x0000000000072be0 ucp_wireup_process_request()  /scrap/jenkins/workspace/ucx-3/contrib/../src/ucp/wireup/wireup.c:430
[2020-04-24T23:02:21.704Z]  2 0x00000000000732ee ucp_wireup_msg_handler()  /scrap/jenkins/workspace/ucx-3/contrib/../src/ucp/wireup/wireup.c:662
[2020-04-24T23:02:21.704Z]  3 0x0000000000025f2d uct_iface_invoke_am()  /scrap/jenkins/workspace/ucx-3/contrib/../src/uct/base/uct_iface.h:632
[2020-04-24T23:02:21.704Z]  4 0x0000000000027c1d uct_ib_iface_invoke_am_desc()  /scrap/jenkins/workspace/ucx-3/contrib/../src/uct/ib/base/ib_iface.h:315
[2020-04-24T23:02:21.704Z]  5 0x0000000000027c1d uct_ud_ep_process_rx()  /scrap/jenkins/workspace/ucx-3/contrib/../src/uct/ib/ud/base/ud_ep.c:693
[2020-04-24T23:02:21.704Z]  6 0x000000000002d332 uct_ud_verbs_iface_poll_rx()  /scrap/jenkins/workspace/ucx-3/contrib/../src/uct/ib/ud/verbs/ud_verbs.c:377
[2020-04-24T23:02:21.704Z]  7 0x000000000002d332 uct_ud_verbs_iface_progress()  /scrap/jenkins/workspace/ucx-3/contrib/../src/uct/ib/ud/verbs/ud_verbs.c:424
[2020-04-24T23:02:21.704Z]  8 0x000000000002b6fb uct_iface_progress()  /scrap/jenkins/workspace/ucx-3/contrib/../src/uct/api/uct.h:3131
[2020-04-24T23:02:21.704Z]  9 0x000000000002bfc8 ucp_worker_iface_check_events_progress()  /scrap/jenkins/workspace/ucx-3/contrib/../src/ucp/core/ucp_worker.c:717
[2020-04-24T23:02:21.704Z] 10 0x000000000003a251 ucs_callbackq_slow_proxy()  /scrap/jenkins/workspace/ucx-3/contrib/../src/ucs/datastruct/callbackq.c:400
[2020-04-24T23:02:21.704Z] 11 0x00000000000279d2 ucs_callbackq_dispatch()  /scrap/jenkins/workspace/ucx-3/contrib/../src/ucs/datastruct/callbackq.h:211
[2020-04-24T23:02:21.704Z] 12 0x0000000000029461 uct_worker_progress()  /scrap/jenkins/workspace/ucx-3/contrib/../src/uct/api/uct.h:2350
[2020-04-24T23:02:21.704Z] 13 0x0000000000764cfc ucp_test_base::entity::progress()  /scrap/jenkins/workspace/ucx-3/contrib/../test/gtest/ucp/ucp_test.cc:749
[2020-04-24T23:02:21.704Z] 14 0x0000000000766af3 ucp_test::progress()  /scrap/jenkins/workspace/ucx-3/contrib/../test/gtest/ucp/ucp_test.cc:152
[2020-04-24T23:02:21.705Z] 15 0x0000000000761de4 test_ucp_sockaddr::one_sided_disconnect()  /scrap/jenkins/workspace/ucx-3/contrib/../test/gtest/ucp/test_ucp_sockaddr.cc:460
[2020-04-24T23:02:21.705Z] 16 0x00000000007488c5 test_ucp_sockaddr_destroy_ep_on_err_onesided_client_cforce_Test::test_body()  /scrap/jenkins/workspace/ucx-3/contrib/../test/gtest/ucp/test_ucp_sockaddr.cc:739
[2020-04-24T23:02:21.705Z] 17 0x000000000054769a ucs::test_base::run()  /scrap/jenkins/workspace/ucx-3/contrib/../test/gtest/common/test.cc:290
[2020-04-24T23:02:21.705Z] 18 0x00000000005476e5 ucs::test_base::TestBodyProxy()  /scrap/jenkins/workspace/ucx-3/contrib/../test/gtest/common/test.cc:316
[2020-04-24T23:02:21.705Z] 19 0x000000000053671b HandleSehExceptionsInMethodIfSupported<testing::Test, void>()  /scrap/jenkins/workspace/ucx-3/contrib/../test/gtest/common/gtest-all.cc:3562
[2020-04-24T23:02:21.705Z] 20 0x000000000052c90b testing::Test::Run()  /scrap/jenkins/workspace/ucx-3/contrib/../test/gtest/common/gtest-all.cc:3635
[2020-04-24T23:02:21.705Z] 21 0x000000000052cbb8 testing::TestInfo::Run()  /scrap/jenkins/workspace/ucx-3/contrib/../test/gtest/common/gtest-all.cc:3812
[2020-04-24T23:02:21.705Z] 22 0x000000000052cd75 testing::TestCase::Run()  /scrap/jenkins/workspace/ucx-3/contrib/../test/gtest/common/gtest-all.cc:3930
[2020-04-24T23:02:21.705Z] 23 0x000000000052cd75 testing::TestCase::total_test_count()  /scrap/jenkins/workspace/ucx-3/contrib/../test/gtest/common/gtest-all.cc:3864
[2020-04-24T23:02:21.705Z] 24 0x000000000052cd75 ???()  /scrap/jenkins/workspace/ucx-3/contrib/../test/gtest/common/gtest-all.cc:3929
[2020-04-24T23:02:21.705Z] 25 0x000000000052d342 testing::internal::UnitTestImpl::RunAllTests()  /scrap/jenkins/workspace/ucx-3/contrib/../test/gtest/common/gtest-all.cc:5808
[2020-04-24T23:02:21.705Z] 26 0x000000000052d342 ???()  /scrap/jenkins/workspace/ucx-3/contrib/../test/gtest/common/gtest-all.cc:5806
[2020-04-24T23:02:21.705Z] 27 0x00000000005362eb HandleSehExceptionsInMethodIfSupported<testing::internal::UnitTestImpl, bool>()  /scrap/jenkins/workspace/ucx-3/contrib/../test/gtest/common/gtest-all.cc:3562
[2020-04-24T23:02:21.705Z] 28 0x000000000052be8a testing::UnitTest::Run()  /scrap/jenkins/workspace/ucx-3/contrib/../test/gtest/common/gtest-all.cc:5422
[2020-04-24T23:02:21.705Z] 29 0x000000000052be8a ???()  /scrap/jenkins/workspace/ucx-3/contrib/../test/gtest/common/gtest-all.cc:5422
[2020-04-24T23:02:21.705Z] 30 0x0000000000536cbd main()  /scrap/jenkins/workspace/ucx-3/contrib/../test/gtest/common/main.cc:102
[2020-04-24T23:02:21.705Z] 31 0x000000347981ed1d __libc_start_main()  ???:0
[2020-04-24T23:02:21.705Z] 32 0x0000000000512cd9 _start()  ???:0
[2020-04-24T23:02:21.705Z] =================================
[2020-04-24T23:02:21.705Z] Sending notification to evgenylek@mellanox.com
[2020-04-24T23:02:24.974Z] [boo31:23906:0:23906] Process frozen...
```

**Comment by evgeny-leksikov** [comment]
@alinask thanks for pointing! the failure should be fixed by https://github.com/openucx/ucx/pull/5058 and actually not related to close protocol
bot:pipe:retest

**Comment by yosefe** [comment]
@evgeny-leksikov need to port fixes from the branch

**Comment by evgeny-leksikov** [comment]
bot:pipe:retest

**Comment by evgeny-leksikov** [comment]
> @evgeny-leksikov need to port fixes from the branch

done

**Pull request merged**